### ISSUE TITLE
inconsistent Boruvka and Prim's MSTs

### ISSUE THREAD

**Issue by azizkayumov** [petabi/petal-clustering#69]
As discussed in #67, there is a small difference between MSTs computed by [Boruvka](https://github.com/petabi/petal-clustering/blob/9d2c1ca0d35033487674cc97f07b94e1c67acc8d/src/hdbscan.rs#L62) and [Prim's](https://github.com/petabi/petal-clustering/blob/9d2c1ca0d35033487674cc97f07b94e1c67acc8d/src/hdbscan.rs#L78) algorithms in HDBSCAN.
Here are the steps to reproduce:
1. Download the [Wine dataset](https://archive.ics.uci.edu/dataset/186/wine+quality).
2. Extract `winequality-white.csv` and put in the project folder. 
3. Replace all `;` with `,` and remove the header info (as required by this lib).
4. Add `println!` to print the MST weight ([this line](https://github.com/petabi/petal-clustering/blob/53c79f0c9f1407bb6aa950190d2aebf9e6612f37/src/hdbscan.rs#L86) seems to be a good place for simplicity):
```
        let mut weight = A::zero();
        for (_, _, w) in &mst {
            weight += *w;
        }
        println!("weight: {:?}", weight);
```
5. Change the [example code](https://github.com/petabi/petal-clustering/blob/53c79f0c9f1407bb6aa950190d2aebf9e6612f37/examples/hdbscan.rs#L41) to disable Boruvka: `boruvka = true` => `boruvka = false`
6. Run the example code:  
`cargo run --example hdbscan winequality-white.csv`
   This prints the following (which is the ground truth exact MST):
```
weight: 26787.419129474838
========= Report =========
# of events processed: 4898
# of features provided: 12
# of clusters: 8
# of events clustered: 1564
# of outliers: 3334
```
7. Change the example code to enable Boruvka: `boruvka = false` => `boruvka = true`:
8. Run the example code again:  
`cargo run --example hdbscan winequality-white.csv`
 This will output:
 ```
weight: 26788.24022864788
========= Report =========
# of events processed: 4898
# of features provided: 12
# of clusters: 8
# of events clustered: 1566
# of outliers: 3332
 ```

A quick fix would be not to use the lower bound function as suggested [here](https://github.com/petabi/petal-clustering/pull/67#issuecomment-1945515835), but this may increase the running time of Boruvka.

Interestingly, [Python HDBSCAN](https://github.com/scikit-learn-contrib/hdbscan/blob/98928d0c095715edc9584e7989bd8559673bc2f0/hdbscan/_hdbscan_boruvka.pyx#L859) does not have this bug if we run the same experiment (with the same configurations). It may be possible that [Ball-tree implementation](https://github.com/petabi/petal-neighbors/blob/main/src/ball_tree.rs) might be causing this (e.g. erroneous lower bounding between tree nodes due to loss of precision?).

**Comment by msk** [comment]
I cannot reproduce with 0.13.0. Do you still see this bug?

**Comment by azizkayumov** [comment]
Yes, I reproduced it somehow, here are the logs for `boruvka=false`:
```
weight: 26787.419129474838
========= Report =========
# of events processed: 4898
# of features provided: 12
# of clusters: 7
# of events clustered: 1564
# of noise events: 3334
# of outliers (prob > 0.8): 76
```
and for `boruvka=true`:
```
weight: 26788.24022864788
========= Report =========
# of events processed: 4898
# of features provided: 12
# of clusters: 7
# of events clustered: 1564
# of noise events: 3334
# of outliers (prob > 0.8): 76
```
